File: txtscrio/show_str.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

#define LOOP_TAIL                                                            \
  if (i == n) /* check string end */                                         \
    break;                                                                   \
                                                                             \
  s++;                                                                       \
  p++;                                                                       \
  i++;

#define LOOP_TAIL_X                                                          \
  LOOP_TAIL                                                                  \
  x++;

void show_str (uint8_t xpos, uint8_t ypos, const String *str, uint8_t attr) {

  uint_least8_t n; // number of input characters left to proceed

  n = Length (str);
  if (n) {
    const unsigned char *s; // input string
    uint16_t *p; // output screen address
    uint_least16_t a; // prepared attribute
    uint_least8_t i; // index for a new character to show (1 based)
    uint_least8_t x, y; // character's screen coordinates

    s = GetString (str);
    a = attr << 8;
    x = xpos;
    y = ypos;
    p = ((uint16_t *) screen_ptr) + (ypos-1)*MaxCol + xpos-1;
    i = 1;

    if ((y < area_y1) || (y > area_y2+1)) {
      while (1) {
        *p = *s | a; // normal

        LOOP_TAIL
      }
    } else if (y == area_y1) {
      while (1) {
        if ((x < area_x1) || (x > area_x2))
          *p = *s | a; // normal

        LOOP_TAIL_X
      }
    } else if (y <= area_y2) {
      uint_least8_t x22 = area_x2+2; // shadow: vertical rectangle's end X

      while (1) {
        if ((x < area_x1) || (x > x22))
          *p = *s | a; // normal
        else if (x > area_x2)
          *((unsigned char *) p) = *s; // shadow

        LOOP_TAIL_X
      }
    } else {
      uint_least8_t x12 = area_x1+2; // shadow: bottom line's start X
      uint_least8_t x22 = area_x2+2; // shadow: bottom line's end X

      while (1) {
        if ((x < x12) && (x > x22))
          *p = *s | a; // normal
        else
          *((unsigned char *) p) = *s; // shadow

        LOOP_TAIL_X
      }
    }
  }
}

#undef LOOP_TAIL
#undef LOOP_TAIL_X

File: txtscrio/pas/show_str.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure show_str(xpos,ypos: Byte; str: String; color: Byte);

var
  s: ^Char; // input string
  p: Pointer; // output screen address
  a: Word; // prepared attribute
  n: Byte; // number of input characters left to proceed
  i: Byte; // index for a new character to show (1-based)
  x12,x21,x22,y11,y21,x,y: Byte;
  c: Char; // character to show

begin
  n := Length (str);
  if (n <> 0) then begin
    s := @str[1];

    x12 := area_x1 + 2; // shadow: bottom line X1
    x21 := area_x2 + 1; // shadow: vertical rectangle X1
    x22 := area_x2 + 2; // shadow: bottom line and vertical rectangle X2
    y11 := area_y1 + 1; // shadow: vertical rectangle Y1
    y21 := area_y2 + 1; // shadow: bottom line Y and vertical rectangle Y2

    i := 1;
    a := color shl 8;
    repeat
      c := s^;
      Inc (s);

      p := Pointer(screen_ptr) + (((ypos-1)*MaxCol + xpos+i-2) shl 1);

      x := xpos+i-1;
      y := ypos;
      if ((x >= x12) and (x <= x22) and (y = y21)
      or  (x >= x21) and (x <= x22) and (y >= y11) and (y <= y21)) then begin
        // in the shadow
        Char(p^) := c;
      end else
      if ((x < area_x1) or (x > area_x2)
      or  (y < area_y1) or (y > area_y2)) then begin
        // outside the area
        Word(p^) := Ord(c) + a;
      end;

      Inc (i);
    until (i > n); // check string end
  end;
end;
